addons/gamification/models/gamification_goal.py

```diff
@@ -0,0 +1,403 @@
+                self.pool['mail.thread']._model.message_post(self.env.cr, self.env.uid, 0, body=body_html, partner_ids=[self.user_id.partner_id.id], context=self.env.context, subtype='mail.mt_comment')
+                return {'to_update': True}
+        return {}
+
+    def _get_write_values(self, new_value):
+        """Generate values to write after recomputation of a goal score"""
+        if new_value == self.current:
+            # avoid useless write if the new value is the same as the old one
+            return {}
+
+        result = {self.id: {'current': new_value}}
+        if (self.definition_id.condition == 'higher' and new_value >= self.target_goal) \
+          or (self.definition_id.condition == 'lower' and new_value <= self.target_goal):
+            # success, do no set closed as can still change
+            result[self.id]['state'] = 'reached'
+
+        elif self.end_date and fields.Date.from_string(fields.Date.today()) > fields.Date.from_string(self.end_date):
+            # check goal failure
+            result[self.id]['state'] = 'failed'
+            result[self.id]['closed'] = True
+
+        return result
+
+    @api.multi
+    def update_goal(self):
+        """Update the goals to recomputes values and change of states
+
+        If a manual goal is not updated for enough time, the user will be
+        reminded to do so (done only once, in 'inprogress' state).
+        If a goal reaches the target value, the status is set to reached
+        If the end date is passed (at least +1 day, time not considered) without
+        the target value being reached, the goal is set as failed."""
+
+        commit = self.env.context.get('commit_gamification', False)
+
+        goals_by_definition = {}
+        for goal in self:
+            goals_by_definition.setdefault(goal.definition_id, []).append(goal)
+
+        for definition, goals in goals_by_definition.items():
+            goals_to_write = dict((goal.id, {}) for goal in goals)
+            if definition.computation_mode == 'manually':
+                for goal in goals:
+                    goals_to_write[goal.id].update(goal._check_remind_delay())
+            elif definition.computation_mode == 'python':
+                # TODO batch execution
+                for goal in goals:
+                    # execute the chosen method
+                    cxt = {
+                        'object': goal,
+                        'env': self.env,
+                        'date': date, 'datetime': datetime, 'timedelta': timedelta, 'time': time
+                    }
+                    code = definition.compute_code.strip()
+                    safe_eval(code, cxt, mode="exec", nocopy=True)
+                    # the result of the evaluated codeis put in the 'result' local variable, propagated to the context
+                    result = cxt.get('result')
+                    if result is not None and type(result) in (float, int, long):
```

Comment:
``` type(result) in (float, int, long) ``` isn't it enough, if result is None then type will be NoneType and it will not pass the truth test